### PR TITLE
4x4 board

### PR DESCRIPTION
**Pull request by RabeaGleissner**
@felipesere @jsuchy

This PR adds the option to play the game on a 4x4 board.
Looking forward to your feedback!

### CODE REVIEW

File: src/main/java/de/rabea/game/Board.java [hunk 1]
```diff
@@ -10,23 +10,34 @@
...
+        this.numberOfCells = gameState.length;
+        this.cells = gameState;
+    }
 
-    public Board() {
-       this(new Mark[] {EMPTY, EMPTY, EMPTY,
-                EMPTY, EMPTY, EMPTY,
-                EMPTY, EMPTY, EMPTY});
+    public Board(int dimension) {
+        this.dimension = dimension;
```
Comment (felipesere) [inline src/main/java/de/rabea/game/Board.java:23]: If you have more than constructor, then they should each other. It's called a 'telescope constructor'.
This should also drop the number instance variables you need.

File: src/main/java/de/rabea/player/UnbeatableComputerPlayer.java [hunk 2]
```diff
@@ -12,23 +12,23 @@ public UnbeatableComputerPlayer(Mark mark) {
 
     @Override
     public int getPosition(Board board) {
-        int remainingMovesCount = board.emptyCells().size();
-        return minimax(remainingMovesCount, Integer.MIN_VALUE, Integer.MAX_VALUE, board, mark).getMove();
+        int MAXIMUM_DEPTH = 8;
```
Comment (felipesere) [inline src/main/java/de/rabea/player/UnbeatableComputerPlayer.java]: Idiomatic Java has camelCase for local variables and UPPER_CASE for constants (`static final`).

File: src/main/java/de/rabea/ui/PrettyBoardPainter.java [hunk 3]
```diff
@@ -7,26 +7,26 @@
...
+            boardImage = printLastPipe(positionNumberToPrint, boardImage, board);
         }
         return boardImage;
     }
 
-    private String printSymbolInCell(int i, String boardImage, Mark cell) {
+    private String printSymbolInCell(int positionNumberToPrint, String boardImage, Mark cell) {
         if (cell == EMPTY) {
-            boardImage += pipe() + i;
+            boardImage += pipe() + String.format("%2d", positionNumberToPrint);
```
Comment (felipesere) [inline src/main/java/de/rabea/ui/PrettyBoardPainter.java]: What does this `Sring.format(...)`? Extract it to a method and give a decent name.

File: src/main/java/de/rabea/game/Board.java [hunk 4]
```diff
@@ -88,9 +99,15 @@ private int emptyCellsCount() {
     }
 
     public int getDimension() {
-        double length = (double) cells().length;
-        double squareRoot = Math.sqrt(length);
-        return (int) squareRoot;
+        return dimension;
+    }
+
+    public boolean is3x3() {
```
Comment (felipesere) [inline src/main/java/de/rabea/game/Board.java]: What do we do if we want to go 5x5? or 2x2? Why don't you expose the size?

File: src/main/java/de/rabea/game/Board.java [hunk 5]
```diff
@@ -101,7 +118,7 @@ public boolean isIndexOfLastCell(int cell) {
         return cell == indexOfLastCell();
     }
 
-    public boolean isIndexOfEndOfFirstOrSecondRow(int cell) {
+    public boolean isEndOfRowIndexExceptLastRow(int cell) {
```
Comment (felipesere) [inline src/main/java/de/rabea/game/Board.java]: This method name feels overly explicit. I feel like there should be a more 'comfortable' way that doesn't need such special-casing...

File: src/test/java/de/rabea/game/BoardTest.java [hunk 6]
```diff
@@ -20,176 +20,298 @@ public void setup() {
...
+    }
+
+    @Test
+    public void listsAllEmptyPositionsFor3x3() {
+        Board board = new Board(3);
+        board.placeMarkOnExistingBoard(1, X);
+        board.placeMarkOnExistingBoard(2, O);
+        board.placeMarkOnExistingBoard(3, X);
+        board.placeMarkOnExistingBoard(6, O);
         List<Integer> emptyCells = Arrays.asList(0, 4, 5, 7, 8);
```
Comment (felipesere) [inline src/test/java/de/rabea/game/BoardTest.java]: `emptyCells` can be inlined and `Arrays.asList(...)` can be statically imported to improve readability.

File: src/test/java/de/rabea/game/BoardTest.java [hunk 7]
```diff
@@ -20,176 +20,298 @@ public void setup() {
...
+    public void isEndOfFirstOrSecondOrThirdRowOn4x4() {
+        Board board = new Board(4);
+        assertTrue(board.isEndOfRowIndexExceptLastRow(3));
+        assertTrue(board.isEndOfRowIndexExceptLastRow(7));
+        assertTrue(board.isEndOfRowIndexExceptLastRow(11));
+        assertFalse(board.isEndOfRowIndexExceptLastRow(15));
+    }
+
+    @Test
+    public void getsAllLinesOf4x4() {
```
Comment (felipesere) [inline src/test/java/de/rabea/game/BoardTest.java]: This test feels excessive.
What feedback did you get from it?

Comment (RabeaGleissner) [inline src/test/java/de/rabea/game/BoardTest.java]: It helped me with updating the methods to get rows/columns/diagonals for a 4x4 board. 

I understand what you mean though. Maybe this is a test that is good for development but can be deleted once the production code works as expected. The methods are still tested through other tests.

**Comment by RabeaGleissner** [comment]
@felipesere Thanks a lot for your comments! Will implement your suggestions.

**Pull request merged**